C7/UIElements/RightClickMenu.cs

```diff
@@ -0,0 +1,174 @@
+		this.Theme = theme;
+
+		this.Hide();
+
+		// Add the menu as a child of "CanvasLayer" to ensure it's drawn overtop of the game map. "CanvasLayer" should have been named
+		// something like "UILayer" since it's the layer that contains all UI elements. "CanvasLayer" is actually its type.
+		game.GetNode("CanvasLayer").AddChild(this);
+	}
+
+	public void Open(Vector2 position)
+	{
+		// Must show the container first in order to update its RectSize
+		this.Show();
+
+		// Move "position" if the menu would extend past the right or bottom edges of the screen
+		Vector2 offScreen = position + this.RectSize - GetViewport().Size;
+		if (offScreen.x > 0) {
+			position.x -= offScreen.x;
+			if (position.x < 0)
+				position.x = 0;
+		}
+		if (offScreen.y > 0) {
+			position.y -= offScreen.y;
+			if (position.y < 0)
+				position.y = 0;
+		}
+
+		this.RectPosition = position;
+	}
+
+	public void CloseAndDelete()
+	{
+		this.QueueFree();
+	}
+
+	private static StyleBoxFlat GetItemStyleBox(Color color)
+	{
+		var styleBox = new StyleBoxFlat();
+		styleBox.BgColor = color;
+		styleBox.ContentMarginLeft   = 4f;
+		styleBox.ContentMarginTop    = 2f;
+		styleBox.ContentMarginRight  = 4f;
+		styleBox.ContentMarginBottom = 2f;
+		return styleBox;
+	}
+
+	public Button AddItem(string text, Texture icon = null)
+	{
+		var button = new Button();
+		button.Text = text;
+		if (icon != null)
+			button.Icon = icon;
+		button.Align = Button.TextAlign.Left;
+		this.AddChild(button);
+		return button;
+	}
+
+	public void RemoveAll()
+	{
+		foreach (Node child in this.GetChildren())
+			child.QueueFree();
+	}
+
+	public override void _Input(InputEvent @event)
+	{
+		bool mouseOverMenu = new Rect2(Vector2.Zero, this.RectSize).HasPoint(this.GetLocalMousePosition()),
+		     escapeKeyWasPressed = (@event is InputEventKey keyEvent) && keyEvent.Pressed && keyEvent.Scancode == (int)Godot.KeyList.Escape,
+		     mouseClickedOutsideMenu = (@event is InputEventMouseButton mouseButtonEvent) && mouseButtonEvent.IsPressed() && ! mouseOverMenu;
+
+		if (escapeKeyWasPressed || mouseClickedOutsideMenu) {
+			this.AcceptEvent(); // Prevents other controls from receiving this event
+			CloseAndDelete();
+
+		// Eat all events other than mouse events while the cursor is over the menu. We want the menu to grab all input while it's open but we
+		// must make sure not to block mouse events from reaching its child buttons. (This had me confused for a while since the Godot docs
+		// say that events reach children before their parents, but the catch is that there are three phases of input processing. The "input"
+		// phase, this function, then "gui input", and finally "unhandled input". If a control eats an event during the "input" phase it won't
+		// proceed to the "gui input" phase where buttons actually respond to it.)
+		} else if (! ((@event is InputEventMouse) && mouseOverMenu))
+			this.AcceptEvent();
+	}
```

Comment:
Interesting... I'll have to refresh myself on the popup logic when I'm less tired and properly compare them, but I see the logic in the approach taken here.